c10/test/util/complex_test_common.h

```diff
@@ -0,0 +1,461 @@
+  test_construct_from_scalar<float>();
+  test_construct_from_scalar<double>();
+
+  static_assert(std::is_convertible<c10::complex<float>, c10::complex<float>>::value, "");
+  static_assert(!std::is_convertible<c10::complex<double>, c10::complex<float>>::value, "");
+  static_assert(std::is_convertible<c10::complex<float>, c10::complex<double>>::value, "");
+  static_assert(std::is_convertible<c10::complex<double>, c10::complex<double>>::value, "");
+
+  static_assert(std::is_constructible<c10::complex<float>, c10::complex<float>>::value, "");
+  static_assert(std::is_constructible<c10::complex<double>, c10::complex<float>>::value, "");
+  static_assert(std::is_constructible<c10::complex<float>, c10::complex<double>>::value, "");
+  static_assert(std::is_constructible<c10::complex<double>, c10::complex<double>>::value, "");
+
+  test_construct_from_other<float, float>();
+  test_construct_from_other<float, double>();
+  test_construct_from_other<double, float>();
+  test_construct_from_other<double, double>();
+}
+
+template<typename scalar_t>
+C10_HOST_DEVICE void test_construct_from_std() {
+  constexpr scalar_t num1 = scalar_t(1.23);
+  constexpr scalar_t num2 = scalar_t(4.56);
+  static_assert(c10::complex<scalar_t>(std::complex<scalar_t>(num1, num2)).real() == num1, "");
+  static_assert(c10::complex<scalar_t>(std::complex<scalar_t>(num1, num2)).imag() == num2, "");
+}
+
+MAYBE_GLOBAL void test_std_conversion() {
+  test_construct_from_std<float>();
+  test_construct_from_std<double>();
+}
+
+#if defined(__CUDACC__) || defined(__HIPCC__)
+template<typename scalar_t>
+void test_construct_from_thrust() {
+  constexpr scalar_t num1 = scalar_t(1.23);
+  constexpr scalar_t num2 = scalar_t(4.56);
+  ASSERT_EQ(c10::complex<scalar_t>(thrust::complex<scalar_t>(num1, num2)).real(), num1);
+  ASSERT_EQ(c10::complex<scalar_t>(thrust::complex<scalar_t>(num1, num2)).imag(), num2);
+}
+#endif
+
+TEST(TestConstructors, FromThrust) {
+#if defined(__CUDACC__) || defined(__HIPCC__)
+  test_construct_from_thrust<float>();
+  test_construct_from_thrust<double>();
+#endif
+}
+
+
+}  // constructors
+
+namespace assignment {
+
+template<typename scalar_t>
+constexpr c10::complex<scalar_t> one() {
+  c10::complex<scalar_t> result(3, 4);
+  result = scalar_t(1);
+  return result;
+}
+
+MAYBE_GLOBAL void test_assign_real() {
+  static_assert(one<float>().real() == float(1), "");
+  static_assert(one<float>().imag() == float(), "");
+  static_assert(one<double>().real() == double(1), "");
+  static_assert(one<double>().imag() == double(), "");
+}
+
+constexpr std::tuple<c10::complex<double>, c10::complex<float>> one_two() {
+  constexpr c10::complex<float> src(1, 2);
+  c10::complex<double> ret0;
+  c10::complex<float> ret1;
+  ret0 = ret1 = src;
+  return std::make_tuple(ret0, ret1);
+}
+
+MAYBE_GLOBAL void test_assign_other() {
+  constexpr auto tup = one_two();
+  static_assert(std::get<c10::complex<double>>(tup).real() == double(1), "");
```

Comment:
type based access for tuples is only valid after C++14 so should we avoid using it?

Reply:
we are already building pytorch with C++14 :)